### PR TITLE
Optimize batch completion

### LINKED ISSUE #379: Using AMQP results in MessageLockLostException
## Who's affected

*  Anyone using Azure Service Bus with AMQP (transport type)

## Symptoms

* `MessageLockLostException` and `NullReferenceException` are thrown during endpoint shutdown

## Original issue

Configuring ASB to use AMQP with transaction mode ReceiveOnly results in a `MessageLockLostException`.

```
2016-10-17 09:01:40.4613|WARN|33|NServiceBus.Transport.AzureServiceBus.MessageReceiverExtensions|A message lock lost exception occured while trying to complete a batch of messages, you may consider to increase the lock duration or reduce the batch size, the exception was The lock supplied is invalid. Either the lock expired, or the message has already been removed from the queue.
Microsoft.ServiceBus.Messaging.MessageLockLostException: The lock supplied is invalid. Either the lock expired, or the message has already been removed from the queue.
   at Microsoft.ServiceBus.Common.AsyncResult.End[TAsyncResult](IAsyncResult result)
   at Microsoft.ServiceBus.Messaging.MessageReceiver.EndCompleteBatch(IAsyncResult result)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transport.AzureServiceBus.MessageReceiverExtensions.<SafeCompleteBatchAsync>d__1.MoveNext()
```

Besides this I got an `NullReferenceException` at endpoint shutdown

```
2016-10-17 09:02:14.0450|WARN|13|SessionExtensions|CloseWithSuppress (System.NullReferenceException): Object reference not set to an instance of an object.
System.NullReferenceException: Object reference not set to an instance of an object.
   at NServiceBus.Transport.AzureServiceBus.MessageReceiverNotifier.<Stop>d__37.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transport.AzureServiceBus.TopologyOperator.<StopNotifiersForAsync>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transport.AzureServiceBus.TopologyOperator.<Stop>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.Transport.AzureServiceBus.MessagePump.<Stop>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.TransportReceiver.<Stop>d__6.MoveNext() in C:\Build\src\NServiceBus.Core\Transports\TransportReceiver.cs:line 57
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.RunningEndpointInstance.<>c.<<StopReceivers>b__8_0>d.MoveNext() in C:\Build\src\NServiceBus.Core\Unicast\RunningEndpointInstance.cs:line 95
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NServiceBus.RunningEndpointInstance.<Stop>d__1.MoveNext() in C:\Build\src\NServiceBus.Core\Unicast\RunningEndpointInstance.cs:line 45
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at SessionExtensions.<CloseWithSuppress>d__1.MoveNext() in C:\Users\ramon\src\particular\EndToEnd\src\PerformanceTests\Utils\SessionExtensions.cs:line 13
```

### PR DESCRIPTION
**Pull request by danielmarbach**
Closes #379 

## TLDR

- This introduces a "reactive" approach to batch completion for receive only mode.
- Makes sure messages are only completed on the receiver they came from
- Reduces the threads spawn when the number of clients are increased
- Autoscaling of completing under heavy load
- Completely reduces the contention on the concurrent stack there was before and numberOfClients thread idling

## MultiProducerConcurrentDispatcher

The multi producer concurrent dispatcher is a reactive collection with slots, batchSize, pushInterval and maximum concurrency. The guarantees are the following:

- It is fully asynchronous
- Pushing to it is thread safe
- It respects the batchSize and automatically fans out if there is more to complete for a given slot when there is still concurrency left
- It completes messages when they arrive slower under the predefined push interval
- Only ever one thread is executing the pump logic but the completions are faned out
- The pump callback will only receive data for a given slot
- No contention between slots
- Lock-free
- Allocates the necessary collections up front for the maxed out scenario. The only allocations you pay during runtime is a Tuple and the async statemachine
- Has FIFO semantics instead of LIFO

untitled diagram

## Configuration used for tests

```
        transport.UseTopology<ForwardingTopology>();
        endpointConfiguration.LimitMessageProcessingConcurrencyTo(128 * 16);
        transport.TransportType(TransportType.NetMessaging);
        transport.Transactions(TransportTransactionMode.ReceiveOnly);
        transport.NumberOfClientsPerEntity(16);
        transport.MessagingFactories().NumberOfMessagingFactoriesPerNamespace(16 * 2).BatchFlushInterval(TimeSpan.FromMilliseconds(100));
        transport.Queues().EnablePartitioning(true).LockDuration(TimeSpan.FromMinutes(1));
        transport.Topics().EnablePartitioning(true);
        transport.MessageReceivers().PrefetchCount(128);
```

## Completion before

[Before.traces.txt](https://github.com/Particular/NServiceBus.AzureServiceBus/files/539451/Before.traces.txt)

Added a trace point to the batch completion to visualize the count and the thread id

```
Count = 1 Thread = 0x5120
Count = 0 Thread = 0x54E4
Count = 0 Thread = 0x2E2C
Count = 0 Thread = 0x3610
Count = 0 Thread = 0x5F08
Count = 0 Thread = 0x619C
Count = 22 Thread = 0x4DF4
Count = 0 Thread = 0x5C98
Count = 0 Thread = 0x6330
Count = 0 Thread = 0x5920
Count = 6 Thread = 0x5A80
Count = 11 Thread = 0x608C
Count = 0 Thread = 0x6330
Count = 0 Thread = 0x619C
Count = 4 Thread = 0x4E80
Count = 3 Thread = 0x5E34
Count = 0 Thread = 0x4298
Count = 0 Thread = 0x4D50
Count = 0 Thread = 0x62A8
```

### Idle time
```
Count = 0 Thread = 0x6228
Count = 0 Thread = 0x63F0
Count = 0 Thread = 0x63F0
Count = 0 Thread = 0x6228
Count = 0 Thread = 0x63F0
Count = 0 Thread = 0x6228
Count = 0 Thread = 0x63F0
Count = 0 Thread = 0x6228
Count = 0 Thread = 0x63F0
Count = 0 Thread = 0x3438
Count = 0 Thread = 0x63F0
Count = 0 Thread = 0x6064
Count = 0 Thread = 0x63FC
Count = 0 Thread = 0x2AA8
Count = 0 Thread = 0x62A4
Count = 0 Thread = 0x2AA8
Count = 0 Thread = 0x6064
Count = 0 Thread = 0x63FC
...
```
## After with new optimizations

Trace with count, slot and thread id in callback

[MPCD.traces.txt](https://github.com/Particular/NServiceBus.AzureServiceBus/files/539458/MPCD.traces.txt)

```
ount = 1 Slot = 3 Thread = 0x45F8
Count = 1 Slot = 4 Thread = 0x45F8
Count = 1 Slot = 6 Thread = 0x45F8
Count = 1 Slot = 8 Thread = 0x45F8
Count = 1 Slot = 9 Thread = 0x45F8
Count = 1 Slot = 11 Thread = 0x45F8
Count = 1 Slot = 12 Thread = 0x45F8
Count = 8 Slot = 0 Thread = 0x5BCC
Count = 7 Slot = 1 Thread = 0x5BCC
Count = 2 Slot = 2 Thread = 0x5BCC
Count = 3 Slot = 4 Thread = 0x5BCC
Count = 1 Slot = 5 Thread = 0x5BCC
Count = 11 Slot = 6 Thread = 0x5BCC
Count = 7 Slot = 7 Thread = 0x5BCC
Count = 5 Slot = 8 Thread = 0x5BCC
Count = 1 Slot = 9 Thread = 0x5BCC
Count = 9 Slot = 10 Thread = 0x5BCC
Count = 2 Slot = 11 Thread = 0x5BCC
Count = 13 Slot = 12 Thread = 0x5BCC
Count = 1 Slot = 14 Thread = 0x5BCC
Count = 1 Slot = 15 Thread = 0x5BCC
Count = 80 Slot = 0 Thread = 0x4A6C
Count = 56 Slot = 1 Thread = 0x4A6C
Count = 70 Slot = 2 Thread = 0x4A6C
Count = 103 Slot = 3 Thread = 0x4A6C
Count = 34 Slot = 4 Thread = 0x4A6C
Count = 63 Slot = 5 Thread = 0x4A6C
Count = 85 Slot = 6 Thread = 0x4A6C
Count = 70 Slot = 7 Thread = 0x4A6C
Count = 17 Slot = 8 Thread = 0x4A6C
Count = 97 Slot = 9 Thread = 0x4A6C
Count = 64 Slot = 10 Thread = 0x4A6C
Count = 58 Slot = 11 Thread = 0x4A6C
Count = 102 Slot = 12 Thread = 0x4A6C
Count = 17 Slot = 13 Thread = 0x4A6C
Count = 98 Slot = 14 Thread = 0x4A6C
Count = 34 Slot = 15 Thread = 0x4A6C
Count = 48 Slot = 0 Thread = 0x4ED4
Count = 72 Slot = 1 Thread = 0x4ED4
Count = 58 Slot = 2 Thread = 0x4ED4
Count = 109 Slot = 3 Thread = 0x4ED4
Count = 93 Slot = 4 Thread = 0x4ED4
Count = 65 Slot = 5 Thread = 0x4ED4
Count = 127 Slot = 6 Thread = 0x4ED4
Count = 58 Slot = 7 Thread = 0x4ED4
Count = 110 Slot = 8 Thread = 0x4ED4
Count = 115 Slot = 9 Thread = 0x4ED4
Count = 64 Slot = 10 Thread = 0x4ED4
Count = 69 Slot = 11 Thread = 0x4ED4
Count = 110 Slot = 12 Thread = 0x4ED4
Count = 111 Slot = 13 Thread = 0x4ED4
Count = 115 Slot = 14 Thread = 0x4ED4
Count = 94 Slot = 15 Thread = 0x4ED4
```

As you can see due to async nature we mostly reuse the same thread for different slots and only things that belong to a slot are completed.

### Idle time

There is nothing ;)

### CODE REVIEW

**Comment by Scooletz** [comment]
I love the tests covering both the MPCD & the async auto reset. It looks that it fails at the transport lvl tests though.

**Comment by SeanFeldman** [comment]
@Scooletz it's not the trasports. TC is choking for another unrelated reason.

File: src/Transport/Utils/AsyncAutoResetEvent.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
+﻿namespace NServiceBus.AzureServiceBus
+{
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Inspired by http://blogs.msdn.com/b/pfxteam/archive/2012/02/11/10266920.aspx
+    /// </summary>
+    class AsyncAutoResetEvent
```
Comment (SeanFeldman) [inline src/Transport/Utils/AsyncAutoResetEvent.cs]: Looks very good. Excellent work @danielmarbach & @Scooletz

**Comment by danielmarbach** [comment]
I think we should tweak the batchSize and the concurrency to something like 

500-1000 and concurrency set to 4-6 which is probably more realistic for scaling since when we have slow transports it doesn't matter we will make sure we stay in the 500 ms time window

**Comment by SeanFeldman** [comment]
This will give us completion capacity of 4K-8K msg/s (completion concurrency of 4) and up to 6K-12 msg/s (completion concurrency of 6) which is very solid. Since we don't have dedicated tasks/thread and we'll only ramp up when needed, let's go with concurrency of 6. For the buffer, I'm leaning to 1K to be on the safe side.

@Particular/azure-service-bus-maintainers chime in

File: src/Transport/Receiving/MessageReceiverNotifier.cs [hunk 2]
```diff
@@ -79,7 +79,7 @@ public void Initialize(EntityInfo entity, Func<IncomingMessageDetails, ReceiveCo
             options.ExceptionReceived += OptionsOnExceptionReceived;
 
             internalReceivers = new IMessageReceiver[numberOfClients];
-            dispatcher = new MultiProducerConcurrentDispatcher<Guid>(5000, TimeSpan.FromMilliseconds(500), 2, numberOfClients);
+            dispatcher = new MultiProducerConcurrentDispatcher<Guid>(1000, TimeSpan.FromMilliseconds(500), 6, numberOfClients);
```
Comment (danielmarbach) [inline src/Transport/Receiving/MessageReceiverNotifier.cs]: @Particular/azure-service-bus-maintainers would it hurt to complete only every second?

**Comment by yvesgoeleven** [comment]
Great change! Maybe one suggestion: Rename MultiProducerConcurrentDispatcher to MultiProducerConcurrentCompletion, otherwise its a bit confusing with the core dispatcher

**Comment by danielmarbach** [comment]
@yvesgoeleven I'll change. Originally developed it outside the ASB repo. In the context of having a dispatcher the name can be confusing

**Comment by danielmarbach** [comment]
@yvesgoeleven pushed the changes including a small cleanup of the callback per receiver

**Comment by SeanFeldman** [comment]
@danielmarbach what about https://github.com/Particular/NServiceBus.AzureServiceBus/pull/384#discussion_r84351399 ?

File: src/Transport/Receiving/MessageReceiverNotifier.cs (hunk 2 above)
Comment (SeanFeldman) [inline src/Transport/Receiving/MessageReceiverNotifier.cs]: 👍

**Comment by danielmarbach** [comment]
The class is designed to be restartable (see tests). Under that light I'd
say no
